sharding/proposer/service_test.go

```diff
@@ -0,0 +1,261 @@
+package proposer
+
+import (
+	"context"
+	"math/big"
+	"testing"
+
+	"crypto/rand"
+
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/accounts"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind/backends"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/sharding"
+	"github.com/ethereum/go-ethereum/sharding/contracts"
+	"gopkg.in/urfave/cli.v1"
+)
+
+var (
+	key, _            = crypto.HexToECDSA("b71c71a67e1177ad4e901695e1b4b9ee17ae16c6668d313eac2f96dbcda3f291")
+	addr              = crypto.PubkeyToAddress(key.PublicKey)
+	accountBalance, _ = new(big.Int).SetString("1001000000000000000000", 10)
+)
+
+// Mock client for testing proposer.
+type mockNode struct {
+	smc         *contracts.SMC
+	t           *testing.T
+	DepositFlag bool
+	backend     *backends.SimulatedBackend
+}
+
+type ChainReader interface {
+	BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error)
+}
+
+func (m *mockNode) BlockByNumber(ctx context.Context, number *big.Int) (*types.Block, error) {
+	return nil, nil
+}
+
+func (m *mockNode) Account() *accounts.Account {
+	return &accounts.Account{Address: addr}
+}
+
+func (m *mockNode) SMCCaller() *contracts.SMCCaller {
+	return &m.smc.SMCCaller
+}
+
+func (m *mockNode) ChainReader() ethereum.ChainReader {
+	return nil
+}
+
+func (m *mockNode) Context() *cli.Context {
+	return nil
+}
+
+func (m *mockNode) Register(s sharding.ServiceConstructor) error {
+	return nil
+}
+
+func (m *mockNode) SMCTransactor() *contracts.SMCTransactor {
+	return &m.smc.SMCTransactor
+}
+
+func (m *mockNode) CreateTXOpts(value *big.Int) (*bind.TransactOpts, error) {
+	txOpts := transactOpts()
+	txOpts.Value = value
+	return txOpts, nil
+}
+
+func (m *mockNode) DepositFlagSet() bool {
+	return m.DepositFlag
+}
+
+func (m *mockNode) DataDirFlag() string {
+	return "/tmp/datadir"
+}
+
+func (m *mockNode) Sign(hash common.Hash) ([]byte, error) {
+	return nil, nil
+}
+
+// Unused mockClient methods.
+func (m *mockNode) Start() error {
+	return nil
+}
+
+func (m *mockNode) Close() {
```

Comment:
Yikes, there are ~75 lines of `mockNode` and many of these are unused.
Is there a better way to do this?
Can we declare an interface for the proposer expected methods?

Take a look at effective go interface definition: https://golang.org/doc/effective_go.html#interfaces
`Interfaces in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here.`

What this means to me: we could compose some interface to explain what a proposer-compatible node looks like.

```golang
type proposeNode interface {
   SMC // SMC interface that has SMCTransactor() and SMCCaller()
   Account() (Account, err)
   CreateTXOpts() (...)
}
```

This ~75 lines of setup with some methods that aren't used by this package make me think that we are not properly making use of interfaces. I.e. The consumer (sharding/proposer) should define the interface that it needs to properly function and the producer (sharding/node) simply provides a full struct. 

Hopefully, that will reduce boilerplate on the test setup and provide clear isolation of responsibilities.

Reply:
This will be addressed with #149